src/vm/dllimportcallback.cpp

```diff
@@ -33,39 +33,83 @@ struct UM2MThunk_Args
     int argLen;
 };
 
-EXTERN_C void STDCALL UM2MThunk_WrapperHelper(void *pThunkArgs,
-                                              int argLen,
-                                              void *pAddr,
-                                              UMEntryThunk *pEntryThunk,
-                                              Thread *pThread);
-
-#ifdef MDA_SUPPORTED
-EXTERN_C void __fastcall CallbackOnCollectedDelegateHelper(UMEntryThunk *pEntryThunk)
+class UMEntryThunkFreeList
 {
-    CONTRACTL
+public:
+    UMEntryThunkFreeList(size_t threshold) :
+        m_threshold(threshold),
+        m_count(0),
+        m_pHead(NULL),
+        m_pTail(NULL),
+        m_crst(CrstUMEntryThunkFreeList)
     {
-        THROWS;
-        GC_TRIGGERS;
-        MODE_COOPERATIVE;
-        SO_TOLERANT;
-        PRECONDITION(CheckPointer(pEntryThunk));
+        WRAPPER_NO_CONTRACT;
     }
-    CONTRACTL_END;
 
-    MdaCallbackOnCollectedDelegate* pProbe = MDA_GET_ASSISTANT(CallbackOnCollectedDelegate);
-    
-    // This MDA must be active if we generated a call to CallbackOnCollectedDelegateHelper
-    _ASSERTE(pProbe);
+    UMEntryThunk *GetUMEntryThunk()
+    {
+        WRAPPER_NO_CONTRACT;
+
+        if (m_count < m_threshold)
+            return NULL;
+
+        CrstHolder ch(&m_crst);
+
+        UMEntryThunk *pThunk = m_pHead;
 
-    if (pEntryThunk->IsCollected())
+        if (pThunk == NULL)
+            return NULL;
+
+        m_pHead = m_pHead->m_pNextFreeThunk;
+        --m_count;
+
+        return pThunk;
+    }
+
+    void AddToList(UMEntryThunk *pThunk)
     {
-        INSTALL_UNWIND_AND_CONTINUE_HANDLER;
-        pProbe->ReportViolation(pEntryThunk->GetMethod());
-        COMPlusThrow(kNullReferenceException);
-        UNINSTALL_UNWIND_AND_CONTINUE_HANDLER;
+        CONTRACTL
+        {
+            NOTHROW;
+        }
+        CONTRACTL_END;
+
+        CrstHolder ch(&m_crst);
+
+        if (m_pHead == NULL)
+        {
+            m_pHead = pThunk;
+            m_pTail = pThunk;
+        }
+        else
+        {
+            m_pTail->m_pNextFreeThunk = pThunk;
+            m_pTail = pThunk;
+        }
+
+        pThunk->m_pNextFreeThunk = NULL;
+
+        ++m_count;
     }
-}
-#endif // MDA_SUPPORTED
+
+private:
+    // Used to delay reusing freed thunks
+    size_t m_threshold;
+    size_t m_count;
+    UMEntryThunk *m_pHead;
+    UMEntryThunk *m_pTail;
+    Crst m_crst;
+};
+
+#define DEFAULT_THUNK_FREE_LIST_THRESHOLD 64
+
+static UMEntryThunkFreeList s_thunkFreeList(DEFAULT_THUNK_FREE_LIST_THRESHOLD);
```

Comment:
This contains `Crst`. `Crst`s in static variables has to be `CrstStatic` to avoid issues like: https://github.com/dotnet/coreclr/issues/13779#issuecomment-328007409

Reply:
Done.